clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -481,24 +517,47 @@ synchronized void resetIdempotentProducerId() {
             throw new IllegalStateException("Cannot reset producer state for a transactional producer. " +
                     "You must either abort the ongoing transaction or reinitialize the transactional producer instead");
         setProducerIdAndEpoch(ProducerIdAndEpoch.NONE);
+        resetSequenceNumbers();
+        transitionTo(State.UNINITIALIZED);
+    }
+
+    private synchronized void resetSequenceForPartition(TopicPartition topicPartition) {
+        topicPartitionBookkeeper.topicPartitionBookkeeping.remove(topicPartition);
+        this.partitionsWithUnresolvedSequences.remove(topicPartition);
+    }
+
+    private synchronized void resetSequenceNumbers() {
         topicPartitionBookkeeper.reset();
         this.partitionsWithUnresolvedSequences.clear();
-        transitionTo(State.UNINITIALIZED);
     }
 
-    synchronized void resetIdempotentProducerIdIfNeeded() {
-        if (!isTransactional()) {
-            if (shouldResetProducerStateAfterResolvingSequences()) {
-                // Check if the previous run expired batches which requires a reset of the producer state.
+    // package-private for unit testing
+    synchronized boolean shouldBumpEpoch() {
+        return epochBumpRequired;
+    }
+
+    synchronized void bumpEpoch() {
```

Comment:
Changed to `bumpIdempotentProducerEpoch` to match `resetIdempotentProducerId`. I thought saying maybe was a bit weird, because in practice we always check whether it's transactional before calling it, so once we call it it definitely will happen